insecure/integration/functional/test/gossipsub/rpc_inspector/validation_inspector_test.go

```diff
@@ -1117,12 +1358,51 @@ func setupTest(t *testing.T, logger zerolog.Logger, role flow.Role, sporkID flow
 		t.Name(),
 		idProvider,
 		p2ptest.WithRole(role),
+		p2ptest.WithGossipSubTracer(meshTracer),
 		internal.WithCorruptGossipSub(corruptlibp2p.CorruptGossipSubFactory(),
 			corruptlibp2p.CorruptGossipSubConfigFactoryWithInspector(corruptInspectorFunc)),
 	)
 	idProvider.On("ByPeerID", victimNode.Host().ID()).Return(&victimIdentity, true).Maybe()
+	idProvider.On("ByPeerID", spammer.SpammerNode.Host().ID()).Return(&spammer.SpammerId, true).Maybe()
+
+	messageIDs := corruptlibp2p.GossipSubMessageIdsFixture(10)
+
+	// create control message with iWant that contains 5 message IDs that were not tracked
+	ctlMsg := spammer.GenerateCtlMessages(int(controlMessageCount), corruptlibp2p.WithIWant(messageCount, messageCount))
+	ctlMsg[0].Iwant[0].MessageIDs = messageIDs
+	// set first 6 message ids to duplicate id > 50% duplicate message ID threshold
+	for i := 0; i <= 6; i++ {
+		ctlMsg[0].Iwant[0].MessageIDs[i] = messageIDs[0]
+	}
 
-	return signalerCtx, cancel, spammer, victimNode, victimIdentity, distributor, validationInspector, idProvider
+	validationInspector.Start(signalerCtx)
+	nodes := []p2p.LibP2PNode{victimNode, spammer.SpammerNode}
+	startNodesAndEnsureConnected(t, signalerCtx, nodes, sporkID)
+	spammer.Start(t)
+	meshTracer.Start(signalerCtx)
+	defer stopTestComponents(t, cancel, nodes, validationInspector, meshTracer)
+
+	// spam the victim with iWant message that contains 60% duplicate message ids more than the configured 50% allowed threshold
+	spammer.SpamControlMessage(t, victimNode, ctlMsg)
+
+	unittest.RequireCloseBefore(t, done, 2*time.Second, "failed to inspect RPC messages on time")
+}
+
+func randomClusterPrefixedTopic() channels.Topic {
+	return channels.Topic(channels.SyncCluster(flow.ChainID(fmt.Sprintf("%d", rand.Uint64()))))
+}
+
+type onNotificationDissemination func(spammer *corruptlibp2p.GossipSubRouterSpammer) func(args mockery.Arguments)
+type mockDistributorOption func(*mockp2p.GossipSubInspectorNotificationDistributor, *corruptlibp2p.GossipSubRouterSpammer)
+
+func withExpectedNotificationDissemination(expectedNumOfTotalNotif int, f onNotificationDissemination) mockDistributorOption {
+	return func(distributor *mockp2p.GossipSubInspectorNotificationDistributor, spammer *corruptlibp2p.GossipSubRouterSpammer) {
+		distributor.
+			On("Distribute", mockery.Anything).
+			Times(expectedNumOfTotalNotif).
+			Run(f(spammer)).
+			Return(nil)
+	}
```

Comment:
should these test helpers be moved to `insecure/integration/functional/test/gossipsub/rpc_inspector/utils.go`?

Reply:
https://github.com/onflow/flow-go/pull/4574/commits/4090cb028094b6a1657c37e30e6753c78c18f6ea